### PR TITLE
[Enhancement][FlatJson] read&write process use type inference

### PR DESCRIPTION
**Pull request by Seaven**
## What I'm doing:
last flat json type inference pr

* json_column_iterator: support read&cast subfield with query type
* json_writer: support subfield type inference

* fix json_exists function bug
* fix get bigint&double from json bug
* fix json function memory leaks

Follow: #42787

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/rowset/json_column_iterator.cpp [hunk 1]
```diff
@@ -86,14 +108,63 @@ Status JsonFlatColumnIterator::init(const ColumnIteratorOptions& opts) {
...
+            ASSIGN_OR_RETURN(auto res, _cast_exprs[i]->evaluate_checked(nullptr, &chunk));
+            auto target = json_column->get_flat_field(i);
+            target->set_delete_state(source->delete_state());
+            if (res->only_null()) {
+                target->append_nulls(source->size());
+            } else if (res->is_constant()) {
+                auto data = down_cast<ConstColumn*>(res.get())->data_column();
+                target->append_value_multiple_times(*data, 0, source->size());
+            } else {
+                target->append(*res, 0, source->size());
```
Comment (stdpain) [inline be/src/storage/rowset/json_column_iterator.cpp:161]: Is it possible that the target and res nullables don't match?

File: be/src/storage/rowset/json_column_iterator.cpp [hunk 2]
```diff
@@ -38,11 +48,14 @@ class JsonFlatColumnIterator final : public ColumnIterator {
 public:
     JsonFlatColumnIterator(ColumnReader* _reader, std::unique_ptr<ColumnIterator>& null_iter,
                            std::vector<std::unique_ptr<ColumnIterator>>& field_iters,
-                           std::vector<std::string>& flat_paths, ColumnAccessPath* path)
+                           std::vector<std::string>& flat_paths, std::vector<LogicalType>& target_types,
+                           std::vector<LogicalType>& source_types, ColumnAccessPath* path)
```
Comment (stdpain) [inline be/src/storage/rowset/json_column_iterator.cpp:52]:
```suggestion
                           std::vector<LogicalType> source_types, ColumnAccessPath* path)
```

File: be/src/storage/rowset/json_column_iterator.cpp [hunk 3]
```diff
@@ -206,8 +267,11 @@ Status JsonFlatColumnIterator::get_row_ranges_by_zone_map(const std::vector<cons
 class JsonDynamicFlatIterator final : public ColumnIterator {
 public:
     JsonDynamicFlatIterator(std::unique_ptr<ScalarColumnIterator>& json_iter, std::vector<std::string>& flat_paths,
-                            ColumnAccessPath* path)
-            : _json_iter(std::move(json_iter)), _flat_paths(std::move(flat_paths)), _path(path){};
+                            std::vector<LogicalType>& target_types, ColumnAccessPath* path)
```
Comment (stdpain) [inline be/src/storage/rowset/json_column_iterator.cpp:270]:
```suggestion
                            std::vector<LogicalType> target_types, ColumnAccessPath* path)
```

File: be/src/storage/rowset/json_column_iterator.cpp (hunk 1 above)
Comment (Seaven) [reply be/src/storage/rowset/json_column_iterator.cpp:161]: I think never, source is a empty column, we can add a dcheck to do that

**Review by stdpain — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**